src/Npgsql/ConnectorPool.cs

```diff
@@ -121,60 +150,71 @@ internal bool TryAllocateFast(NpgsqlConnection conn, [NotNullWhen(true)] out Npg
                     return connector;
                 }
 
-                break;
-            }
-
-            // We're at max capacity. Asynchronously block on the idle channel with a timeout.
+                // We're at max capacity. Asynchronously block on the idle channel with a timeout.
 
-            // TODO: Potential issue: we only check to create new connections once above. In theory we could have
-            // many attempts waiting on the idle channel forever, since all connections were broken by some network
-            // event. Pretty sure this issue exists in the old lock-free implementation too, think about it (it would
-            // be enough to retry the physical creation above from time to time).
+                // TODO: Potential issue: we only check to create new connections once above. In theory we could have
```

Comment:
Not sure what past @roji had in mind exactly, but I think it's this: if an attempt is waiting in the ReadAsync loop below, and a broken connector is returned to the pool, then the attempt should retry creating a new connector (since total connectors has gone down), but right now we're stuck in an endless loop instead. Seems pretty trivial to fix.

Reply:
Fixed this.